spatial/kdtree/kdtree.go

```diff
@@ -0,0 +1,465 @@
+
+// Dim is an index into a point's coordinates.
+type Dim int
+
+// Comparable is the element interface for values stored in a k-d tree.
+type Comparable interface {
+	// Compare returns the shortest translation of the plane through b with
+	// normal vector along dimension d to the parallel plane through a.
+	//
+	// Given c = a.Compare(b, d):
+	//  c = a_d - b_d
+	//
+	Compare(Comparable, Dim) float64
+
+	// Dims returns the number of dimensions described in the Comparable.
+	Dims() int
+
+	// Distance returns the squared Euclidean distance between the receiver and
+	// the parameter.
+	Distance(Comparable) float64
+}
+
+// Extender is a Comparable that can increase a bounding volume to include the
+// point represented by the Comparable.
+type Extender interface {
+	Comparable
+
+	// Extend returns a bounding box that has been extended to include the
+	// receiver. Extend may return nil.
+	Extend(*Bounding) *Bounding
+}
+
+// Bounding represents a volume bounding box.
+type Bounding struct {
+	Min, Max Comparable
+}
+
+// Contains returns whether c is within the volume of the Bounding. A nil Bounding
+// returns true.
+func (b *Bounding) Contains(c Comparable) bool {
+	if b == nil {
+		return true
+	}
+	for d := Dim(0); d < Dim(c.Dims()); d++ {
+		if c.Compare(b.Min, d) < 0 || 0 < c.Compare(b.Max, d) {
+			return false
+		}
+	}
+	return true
+}
+
+// Node holds a single point value in a k-d tree.
+type Node struct {
+	Point       Comparable
+	Plane       Dim
+	Left, Right *Node
+	*Bounding
+}
+
+func (n *Node) String() string {
+	if n == nil {
+		return "<nil>"
+	}
+	return fmt.Sprintf("%.3f %d", n.Point, n.Plane)
+}
+
+// Tree implements a k-d tree creation and nearest neighbor search.
+type Tree struct {
+	Root  *Node
+	Count int
+}
+
+// New returns a k-d tree constructed from the values in p. If p is a Bounder and
+// bounding is true, bounds are determined for each node.
+func New(p Interface, bounding bool) *Tree {
+	if p, ok := p.(bounder); ok && bounding {
+		return &Tree{
+			Root:  buildBounded(p, 0, bounding),
+			Count: p.Len(),
+		}
+	}
+	return &Tree{
+		Root:  build(p, 0),
+		Count: p.Len(),
+	}
+}
+
+func build(p Interface, plane Dim) *Node {
+	if p.Len() == 0 {
+		return nil
+	}
+
+	piv := p.Pivot(plane)
+	d := p.Index(piv)
+	np := (plane + 1) % Dim(d.Dims())
+
+	return &Node{
+		Point:    d,
+		Plane:    plane,
+		Left:     build(p.Slice(0, piv), np),
+		Right:    build(p.Slice(piv+1, p.Len()), np),
+		Bounding: nil,
+	}
+}
+
+func buildBounded(p bounder, plane Dim, bounding bool) *Node {
+	if p.Len() == 0 {
+		return nil
+	}
+
+	piv := p.Pivot(plane)
+	d := p.Index(piv)
+	np := (plane + 1) % Dim(d.Dims())
+
+	b := p.Bounds()
+	return &Node{
+		Point:    d,
+		Plane:    plane,
+		Left:     buildBounded(p.Slice(0, piv).(bounder), np, bounding),
+		Right:    buildBounded(p.Slice(piv+1, p.Len()).(bounder), np, bounding),
+		Bounding: b,
+	}
+}
+
+// Insert adds a point to the tree, updating the bounding volumes if bounding is
```

Comment:
After several readings I can match the comment to the code but it did take me some time. I don't have an alternative suggestion but does `Insert` have to take `bounding bool`? Can't the `bounding` passed to `New` determine the behavior of the whole tree?

Reply:
It's possible to build a tree without using `New`, so unfortunately it's not possible to know what the intention is for the first point being added without having this parameter.